DRAFT: Fix/environment loader

pquerner opened the pull request:
<!---
    Thank you for contributing to OpenMage LTS.
    To help us process this pull request we recommend that you add the following information:
     - Summary of the pull request,
     - Issue(s) related to the changes made,
     - Manual testing scenarios
    Fields marked with (*) are required. Please don't remove the template.
-->

<!--- Please provide a general summary of the Pull Request in the Title above -->

### Description (*)
<!---
    Please provide a description of the changes proposed in the pull request.
    Letting us know what has changed and why it needed changing will help us validate this pull request.
-->
This PR attempts to fix the incomplete attempt for config values to be overwritten from ENV variables.

### Related Pull Requests
<!-- related pull request placeholder -->

### Fixed Issues (if relevant)
<!---
    If relevant, please provide a list of fixed issues in the format OpenMage/magento-lts#<issue_number>.
    There could be 1 or more issues linked here and it will help us find some more information about the reasoning behind this change.
-->
Relates to issue https://github.com/OpenMage/magento-lts/issues/4257#issuecomment-2599524399

### Manual testing scenarios (*)
<!---
    Please provide a set of unambiguous steps to test the proposed code change.
    Giving us manual testing scenarios will help with the processing and validation process.
-->
1. Variables from the ENV should now be disabled for editing on the backend, but still shown.
2. Variables from the ENV should be available from `Mage::getStoreConfig`.

### Questions or comments
<!---
	If relevant, here you can ask questions or provide comments on your pull request for the reviewer
	For example if you need assistance with writing tests or would like some feedback on one of your development ideas
-->
Please give feedback to this WIP solution. 
Maybe in the future we can add a reasoning of "why" the field in the backend is disabled for edits. Such as give a "warning" sign and show that this field is overwritten by a environment variable.
I feel that this is out of scope for now, for I think that the feature should first work the best it can.

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All automated tests passed successfully (all builds are green)

pquerner commented:
@sreichel Please see this PR.
(From your comment on https://github.com/OpenMage/magento-lts/issues/4257#issuecomment-2599524399)

sreichel commented on app/code/core/Mage/Adminhtml/Block/System/Config/Form.php:
```diff
@@ -629,6 +630,20 @@ public function getScope()
         return $scope;
     }
 
+    /**
+     * Render element as disabled, if overwritten by ENV variable
+     *
+     * @param string $data
+     * @return bool
+     */
+    public function isDisabled($path): bool
```
```suggestion
    /**
     * Render element as disabled, if overwritten by ENV variable
     */
    public function isDisabled(string $path): bool
```

sreichel commented on app/code/core/Mage/Core/Helper/EnvironmentConfigLoader.php:
```diff
@@ -70,18 +70,106 @@ public function overrideEnvironment(Varien_Simplexml_Config $xmlConfig)
                     break;
             }
         }
+        $hasConfig = array_key_exists($wantedPath, $config);
+        return $hasConfig;
```
```suggestion
        return array_key_exists($wantedPath, $config);
```

sreichel commented:
Quick test. 404 one every page.

sreichel commented:
Maybe it come from invalid setting .... ?

```
web_environment:
    - MAGE_IS_DEVELOPER_MODE=1
    - OPENMAGE_CONFIG__DEFAULT__GENERAL__STORE_INFORMATION__NAME=ENV default
    - OPENMAGE_CONFIG__DEFAULT__GENERAL__FOO-BAR__NAME=ENV default dashes
    - OPENMAGE_CONFIG__DEFAULT__GENERAL__FOO_BAR__NAME=ENV default underscore
    - OPENMAGE_CONFIG__DEFAULT__GENERAL__ST=ENV default invalid
    - OPENMAGE_CONFIG__WEBSITES__BASE__GENERAL__STORE_INFORMATION__NAME=ENV website
    - OPENMAGE_CONFIG__WEBSITES__BASE-AT__GENERAL__STORE_INFORMATION__NAME=ENV website dashes
    - OPENMAGE_CONFIG__WEBSITES__BASE_CH__GENERAL__STORE_INFORMATION__NAME=ENV website underscore
    - OPENMAGE_CONFIG__WEBSITES__BASE__GENERAL__ST=ENV website invalid
    - OPENMAGE_CONFIG__STORES__GERMAN__GENERAL__STORE_INFORMATION__NAME=ENV store
    - OPENMAGE_CONFIG__STORES__GERMAN-AT__GENERAL__STORE_INFORMATION__NAME=ENV store dashes
    - OPENMAGE_CONFIG__STORES__GERMAN_CH__GENERAL__STORE_INFORMATION__NAME=ENV store underscore
    - OPENMAGE_CONFIG__STORES__GERMAN__GENERAL__ST=ENV store invalid
```

pquerner commented:
For reference my old config:

```
#web_environment: [
#    "OPENMAGE_CONFIG__DEFAULT__GENERAL__STORE_INFORMATION__NAME=asd_env",
#    "OPENMAGE_CONFIG__WEBSITES__BASE__GENERAL__STORE_INFORMATION__NAME=website_env",
#    "OPENMAGE_CONFIG__STORES__GERMAN__GENERAL__STORE_INFORMATION__NAME=store_german_env"
#]
```

With yours I am getting infinite loop.
[infinteloop.html.zip](https://github.com/user-attachments/files/18723627/infinteloop.html.zip)

pquerner commented:
The error was caused by the `Mage::logException` call - which finds the file via `getConfigStore` - which creates an infinite loop.
I've added `Mage::registry`  checks, so that these methods dont run more than needed and removed the `logException` calls to counter this error.

via https://github.com/OpenMage/magento-lts/pull/4617/commits/1f6b7ed9286da386e38d905168e35a99f32c222b

sreichel commented:
Thanks for working on it.

Cant make a PR to your repo right now ... how about to filter env vars in getEnv()?

```
    public function getEnv(): array
    {
        if (is_null($this->envStore)) {
            $env = getenv();
            $env = array_filter($env, function ($key) {
                return str_starts_with($key, self::ENV_STARTS_WITH);
            }, ARRAY_FILTER_USE_KEY);

            $this->envStore = $env;
        }
        return $this->envStore;
    }
```

Maybe add an additional check vor env `OPENMAG_CONFIG_OVERRIDE_ALLOWED` to enable/disable that feature?

sreichel commented:
> 1. Variables from the ENV should now be disabled for editing on the backend, but still shown.

Yes, on default scope, but on one website/storevie level.

(Scope label should ben changed to "ENV")

pquerner commented:
> > 1. Variables from the ENV should now be disabled for editing on the backend, but still shown.
> 
> Yes, on default scope, but not on one website/storeview level.
> 
> (Scope label should ben changed to "ENV")

What do you mean by that?

<img width="1037" alt="image" src="https://github.com/user-attachments/assets/c0bd9c41-8af4-4dfa-a6f4-93ab276baec3" />

You want "Store Name" to be "ENV Store Name"?

All your other suggestions have been added.

Thanks. :)

sreichel commented:
> What do you mean by that?

In "Default" scope the attribute is locked. For "German" storeview its not.

> You want "Store Name" to be "ENV Store Name"?

No, not that other "label". :) "[STORE VIEW]" could change to "[ENV]" and the checkbox can be removed. 

I am busy for the next 2-3 days. I'll test ASAP.

pquerner commented:
> In "Default" scope the attribute is locked. For "German" storeview its not.

Hm. For me its locked. 

> No, not that other "label". :) "[STORE VIEW]" could change to "[ENV]" and the checkbox can be removed.

Gotcha. Good point.

sreichel commented:
> Hm. For me its locked.

Not sure about it, sorry. Maybe the disabled checkbox was missleading. (?)

pquerner commented:
Its now this:

<img width="1083" alt="image" src="https://github.com/user-attachments/assets/8076bffe-160c-4516-ae97-e48d395aee8a" />

sreichel commented:
Great! :heart: 

I can do it later, but if you have time for it, please check the red SonarCloud issues.

pquerner commented:
> Great! ❤️
> 
> I can do it later, but if you have time for it, please check the red SonarCloud issues.

I've removed one case.
The last ones are: unused variable (from the `list`) and the use of Reflection class.

I'm not sure how else to get there (without makeing the cache array public?)

sreichel commented:
To pass the complexity-check it should be enough to move try/catch-parts to a seperate method.

Leave it as it is, i'll add a PR later.

PS: please `composer run php-cs-fixer:fix` .... :P

pquerner commented:
```
ddev php-cs-fixer
Loaded config default from "/var/www/html/.php-cs-fixer.dist.php".

In RuleSets.php line 65:
                                    
  Set "@PER-CS2.0" does not exist.
```

:/

sreichel commented on tests/unit/Mage/Core/Helper/EnvironmentConfigLoaderTest.php:
```diff
@@ -53,6 +56,30 @@ public function testBuildPath(): void
+    {
+        $environmentConfigLoaderHelper = new EnvironmentConfigLoaderTestHelper();
+        $environmentConfigLoaderHelper->setEnvStore([
+            'OPENMAGE_CONFIG__DEFAULT__GENERAL__STORE_INFORMATION__NAME' => "some_value"
```
```suggestion
            'OPENMAGE_CONFIG__DEFAULT__GENERAL__STORE_INFORMATION__NAME' => 'some_value'
```

sreichel commented on tests/unit/Mage/Core/Helper/EnvironmentConfigLoaderTest.php:
```diff
@@ -53,6 +56,30 @@ public function testBuildPath(): void
+        $this->assertIsArray($env);
+        $this->assertEmpty($env);
+        $environmentConfigLoaderHelper->setEnvStore([
+            'OPENMAGE_CONFIG__DEFAULT__GENERAL__STORE_INFORMATION__NAME' => "some_value",
+            'OPENMAGE_CONFIG_OVERRIDE_ALLOWED' => 1 // enable feature
```
```suggestion
            'OPENMAGE_CONFIG__DEFAULT__GENERAL__STORE_INFORMATION__NAME' => 'some_value',
            'OPENMAGE_CONFIG_OVERRIDE_ALLOWED' => 1, // enable feature
```

sreichel commented on app/code/core/Mage/Core/Helper/EnvironmentConfigLoader.php:
```diff
@@ -70,18 +75,116 @@ public function overrideEnvironment(Varien_Simplexml_Config $xmlConfig)
+                    break;
+            }
+        }
+        Mage::register("current_env_config", true, true);
```
```suggestion
        Mage::register('current_env_config', true, true);
```

sreichel commented on app/code/core/Mage/Adminhtml/Block/System/Config/Form.php:
```diff
@@ -389,9 +391,16 @@ public function initFields($fieldset, $group, $section, $fieldPrefix = '', $labe
-                    'can_use_default_value' => $this->canUseDefaultValue((int) $element->show_in_default),
-                    'can_use_website_value' => $this->canUseWebsiteValue((int) $element->show_in_website),
-                ]);
+                    'can_use_default_value' => $this->canUseDefaultValue((int)$element->show_in_default),
+                    'can_use_website_value' => $this->canUseWebsiteValue((int)$element->show_in_website),
```
```suggestion
                    'can_use_default_value' => $this->canUseDefaultValue((int) $element->show_in_default),
                    'can_use_website_value' => $this->canUseWebsiteValue((int) $element->show_in_website),
```

sreichel commented on tests/unit/Mage/Core/Helper/EnvironmentConfigLoaderTest.php:
```diff
@@ -53,6 +56,30 @@ public function testBuildPath(): void
+    {
+        $environmentConfigLoaderHelper = new EnvironmentConfigLoaderTestHelper();
+        $environmentConfigLoaderHelper->setEnvStore([
+            'OPENMAGE_CONFIG__DEFAULT__GENERAL__STORE_INFORMATION__NAME' => 'some_value'
```
```suggestion
            'OPENMAGE_CONFIG__DEFAULT__GENERAL__STORE_INFORMATION__NAME' => 'some_value',
```

sreichel replied:
Last one.

sreichel commented on app/code/core/Mage/Adminhtml/Block/System/Config/Form.php:
```diff
@@ -629,6 +638,18 @@ public function getScope()
     }
 
+    /**
+     * Returns true if element was overwritten by ENV variable
+     * @return bool
```
```suggestion
     * Returns true if element was overwritten by ENV variable
```

Rector

pquerner commented:
Manually tested upgrade scripts with this:

```
<?php

/** @var Mage_Core_Model_Resource_Setup $installer */
$installer = $this;
$installer->startSetup();

$f = Mage::getStoreConfig('general/store_information/name');
$f2 = Mage::getStoreConfig('general/store_information/name', 'store_german');

Mage::getSingleton('core/config')
    ->saveConfig('general/store_information/name', 'test', 'stores', '2');

$installer->endSetup();

```

Values were fine. The insert to db worked, while that entry should've a env value.

Calling `Mage::getStoreConfig` on it yields the `database` value - which is wrong. I'll investigate.

sreichel commented:
Looks (nearly?) complete :)

Ready to review?

pquerner commented:
Yes, unless you can think of some more tests maybe?
I think a automated test for `getStoreConfig` couldnt hurt, although I am not sure how yet. The current tests `should've` done that - but eh.. here we are. :D

sreichel commented:
> Yes, unless you can think of some more tests maybe?

Hehe, thats may the only thing ... have you tested code-coverage for all new methods?

> I think a automated test for `getStoreConfig` couldnt hurt, although I am not sure how yet.

Idea only ....

- add ENV vars to phpunit config
- set OVERRIDE_ALLOWED in test via putenv()
- assert ENV-value

Will test as soon i am back.

sreichel commented:
Backend looks good, but `Mage::getStoreConfig('general/store_information/name')` returns an empty string.

pquerner commented:
> Backend looks good, but `Mage::getStoreConfig('general/store_information/name')` returns an empty string. (sample datas default)

From which context?

For me its working. Doesnt matter if the db has existing value or not. Or frontend or adminhtml (controller, setup scripts) context.

sreichel commented:
> From which context? Which file did you put it?

A standalone file that i use for simple debugging ...

```php
<?php

declare(strict_types=0);

require_once('app/Mage.php');
umask(0);
Mage::app();

echo '<pre>';

# ....

function test4617()
{
    var_dump(Mage::getStoreConfig('general/store_information/name'));
}

$test = $_GET['test'];
if (is_string($test) && function_exists($test)) {
    $test();
}

```

pquerner commented:
Hm somethings not right on your end. Even with your test script I am getting the env variable. 🤷

sreichel commented:
Working on php81 branch right now ... can you please test ith with that config (https://github.com/OpenMage/magento-lts/pull/4617#issuecomment-2645980023)?

Also faced some issues (need to verify that) in admin config sections - like missing dropdowns. (Possibly from another change i made?)

pquerner commented:
> ... can you please test ith with that config (https://github.com/OpenMage/magento-lts/pull/4617#issuecomment-2645980023)?

That is the config I am using right now.

sreichel commented:
Only an idea ... is config cache possibly disabled?

pquerner commented:
All my cached are `enabled`. But even if they are `disabled` it works. :D

sreichel commented:
Okay okay :D I'll test it later.

pquerner commented:
`like missing dropdowns.` thats interesting.
I have overwritten the dropdown option: `OPENMAGE_CONFIG__DEFAULT__GENERAL__STORE_INFORMATION__MERCHANT_COUNTRY=DE` and while its working in adminhtml, it isnt for `getStoreConfig` calls.

sreichel commented:
Seems you could reproce my issue :+1:

pquerner commented:
> Seems you could reproduce my issue 👍

Not with the `general/store_information/name`. `general/store_information/merchant_country` is completely new (and wasnt in your original config, nor mine.)

pquerner commented:
I found it not working and I need to work on it some more.

sreichel commented:
For rector you can try "composer run rector:fix"

pquerner commented:
Nice! What are the next steps?

sreichel commented on tests/unit/Mage/Core/Helper/EnvironmentConfigLoaderTest.php:
```diff
@@ -170,6 +205,104 @@ public function envOverridesCorrectConfigKeysDataProvider(): Generator
+                'env_path'  => 'OPENMAGE_CONFIG__DEFAULT__GENERAL__STORE_INFORMATION__NAME',
+                'store'  => '', // or 'default', which will be used internally, but this is how \Mage_Adminhtml_Model_Config_Data::_validate defines it
+                'expected'  => [
+                    'general/store_information/name' => 1,
```
```suggestion
                    self::XML_PATH_GENERAL => 1,
```

sreichel commented on tests/unit/Mage/Core/Helper/EnvironmentConfigLoaderTest.php:
```diff
@@ -170,6 +205,104 @@ public function envOverridesCorrectConfigKeysDataProvider(): Generator
+                'env_path'  => 'OPENMAGE_CONFIG__STORES__GERMAN__GENERAL__STORE_INFORMATION__NAME',
+                'store'  => 'german',
+                'expected'  => [
+                    'general/store_information/name' => 1,
```
```suggestion
                    self::XML_PATH_GENERAL => 1,
```

sreichel commented:
Docs? In a new PR?

Made a small PR at your repo to fix some errors reported by sonar.

sreichel commented:
Ready for review?

sreichel commented:
Please merge https://github.com/pquerner/magento-lts/pull/2

pquerner commented:
@sreichel Merged your PR and fixed PHPCS. I am having trouble with rector. Could you please take a look?

pquerner commented:
> @pquerner what problems do you face with rector? `composer run rector:fix` should be okay.

https://github.com/OpenMage/magento-lts/actions/runs/18821943379/job/53698956967?pr=4617
They fail

sreichel commented:
## Tests

### Code `Mage::getStoreConfig`

- [x] passing `null` as store-code returns empty string
- [x] DEFAULT env scope does not overwrite website/store config
- [x] WEBSITE env scope does not overwrite store config

### Admin backend
- [x] DEFAULT env scope does not overwrite all websites/stores config
- [x] WEBSITE env scope does not overwrite alls related stores config
- [x]  locked configs wrong

### ToDo

- [ ] reduce calls to Mage::getStores() ... another PR?

sreichel commented:
Thinks its ready to review?

pquerner commented:
Website dash (-) can be removed, yes. Tests fail because they iterate through the stores and no stores with german-at or german_ch exists (anymore?).
Idk need to look into it.

sreichel commented:
Merged.

This feature needs to be enabled via ENV `OPENMAGE_CONFIG_OVERRIDE_ALLOWED`.
